### ISSUE TITLE
Android app crash

### ISSUE THREAD

**Issue by pesseba** [khindemit/unity-project-vorbis#1]
I build the project in Unity for an Android app, The app starts fine, shows the logo intro, but when the scene is open (I can see the UI for a moment), the app just closes. 
I tried to see some log error in abdTool, but there is no errors there... 
I generated the lib 'libVorbisPlugin.so' in Android Studio. I found onlyt the armeabi-v7a version. Could be this?
How can I generate the other versions?

**Comment by gindemit** [comment]
Hey pesseba!
I've created the first release with the builded libraries:
https://github.com/khindemit/unity-vorbis/releases/tag/0.1

You can find also the versions for ARM64

**Comment by pesseba** [comment]
Thank you @khindemit ! 

I tried to run the Example Project in Editor, but I got some errors. I checked the plataform settings of the plugins, its right.
I build an Android version too. The app demo runs normally. 
Without the windows version working in editor mode, its hard to keep a development environment.

The error is about some dependency, as follow:

```
(Filename: Assets/_Source/PluginTest.cs Line: 42)

Plugins: Failed to load 'Assets/Plugins/Windows/x86_64/VorbisPlugin.dll' because one or more of its dependencies could not be loaded.
 
(Filename: C:\buildslave\unity\build\Runtime/Misc/Plugins.cpp Line: 292)

Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
Plugins: Failed to load 'Assets/Plugins/Windows/x86_64/VorbisPlugin.dll' because one or more of its dependencies could not be loaded.
StreamTest:OnEnable() (at Assets\_Source\StreamTest.cs:13)

[C:\buildslave\unity\build\Runtime/Misc/Plugins.cpp line 292] 
(Filename: Assets/_Source/StreamTest.cs Line: 13)

Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
DllNotFoundException: VorbisPlugin
  at (wrapper managed-to-native) VorbisPlugin.OpenReadFileStream(string,int16&,int&)
  at StreamTest.OnEnable () [0x0001d] in C:\Users\Rodrigo\Downloads\unity-project-vorbis-master\UnityClient\Assets\_Source\StreamTest.cs:13 
 
(Filename: Assets/_Source/StreamTest.cs Line: 13)

Loaded scene 'Temp/__Backupscenes/0.backup'
	Deserialize:            445.089 ms
	Integration:            356.577 ms
	Integration of assets:  13.499 ms
	Thread Wait Time:       18.882 ms
	Total Operation Time:   834.047 ms
Plugins: Failed to load 'Assets/Plugins/Windows/x86_64/VorbisPlugin.dll' because one or more of its dependencies could not be loaded.
 
(Filename: C:\buildslave\unity\build\Runtime/Misc/Plugins.cpp Line: 292)

Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
Plugins: Failed to load 'Assets/Plugins/Windows/x86_64/VorbisPlugin.dll' because one or more of its dependencies could not be loaded.
StreamTest:OnAudioFilterRead(Single[], Int32) (at Assets\_Source\StreamTest.cs:25)

[C:\buildslave\unity\build\Runtime/Misc/Plugins.cpp line 292] 
(Filename: Assets/_Source/StreamTest.cs Line: 25)

Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/VorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin.dll
Fallback handler could not load library C:/Program Files/Unity/Hub/Editor/2019.2.18f1/Editor/Data/Mono/lib/libVorbisPlugin
DllNotFoundException: VorbisPlugin
  at (wrapper managed-to-native) VorbisPlugin.ReadFromFileStream(intptr,single[],int)
  at StreamTest.OnAudioFilterRead (System.Single[] data, System.Int32 channels) [0x00007] in C:\Users\Rodrigo\Downloads\unity-project-vorbis-master\UnityClient\Assets\_Source\StreamTest.cs:25 
 
(Filename: Assets/_Source/StreamTest.cs Line: 25)

DllNotFoundException: VorbisPlugin
  at (wrapper managed-to-native) VorbisPlugin.ReadFromFileStream(intptr,single[],int)
  at StreamTest.OnAudioFilterRead (System.Single[] data, System.Int32 channels) [0x00007] in C:\Users\Rodrigo\Downloads\unity-project-vorbis-master\UnityClient\Assets\_Source\StreamTest.cs:25 
 
(Filename: Assets/_Source/StreamTest.cs Line: 25)

DllNotFoundException: VorbisPlugin
  at (wrapper managed-to-native) VorbisPlugin.ReadFromFileStream(intptr,single[],int)
  at StreamTest.OnAudioFilterRead (System.Single[] data, System.Int32 channels) [0x00007] in C:\Users\Rodrigo\Downloads\unity-project-vorbis-master\UnityClient\Assets\_Source\StreamTest.cs:25 
 
(Filename: Assets/_Source/StreamTest.cs Line: 25)
```

**Comment by pesseba** [comment]
Hi @khindemit,
I run abdTool in Android version, and it got some error too...
The app dispatch a loop of this error:

```
11-20 08:59:05.882 21915 21961 E Unity   : EntryPointNotFoundException: ReadFromFileStream
11-20 08:59:05.882 21915 21961 E Unity   :   at (wrapper managed-to-native) VorbisPlugin.ReadFromFileStream(intptr,single[],int)
11-20 08:59:05.882 21915 21961 E Unity   :   at StreamTest.OnAudioFilterRead (System.Single[] data, System.Int32 channels) [0x00006] in <fda8cb5f8af844cc844b66ed9f8b8215>:0
```
You can download the app here: https://we.tl/t-oFCqQN08Tp

**Comment by ChristianMcFarlaneWork** [comment]
Hey all,
Just came across this issue and wanted to follow up in case anyone else has the same problem that I've had with the issue and is looking for a possible solution.

After failing to build the libraries directly from Android Studio, I downloaded the built libraries provided above and was receiving a similar error of:
```
System.EntryPointNotFoundException: Unable to find an entry point named 'WriteAllPcmDataToFile' in 'VorbisPlugin'.
  at VorbisPlugin.WriteAllPcmDataToFile (System.String filePath, System.Single[] samples, System.Int32 samplesLength, System.Int16 channels, System.Int32 frequency, System.Single base_quality, System.Int32 samplesToRead) [0x00000] in <00000000000000000000000000000000>:0 
```
After looking at a bunch of other similar issues and assuming there was some kind of name mangling going on, I had a look at the Symbols of the '.so' using objdump and it seems that in the prebuilt assemblies provided [above](https://github.com/khindemit/unity-project-vorbis/issues/1#issuecomment-730634738) use different names for the importing and exporting functions.

### What fixed the issue for me
The fix for this issue is to open up 'VorbisPlugin.cs' and change the function names for both 
`WriteAllPcmDataToFile` and `ReadAllPcmDataFromFile` into `EncodePcmDataToFile` and `DecodePcmDataFromFile` respectively. so you end up with 
```
    [DllImport(PLUGIN_NAME)]
    private static extern long EncodePcmDataToFile(string filePath, float[] samples, int samplesLength, short channels, int frequency, float base_quality, int samplesToRead);

    [DllImport(PLUGIN_NAME)]
    private static extern long DecodePcmDataFromFile(string filePath, out System.IntPtr samples, out int samplesLength, out short channels, out int frequency, int maxSamplesToRead);
```

This solved the main issue of the plugin crashing and not being able to find the entry point. Hope this helps someone else a number of hours of frustration.

Now I just have to figure out why it's saving the audio at half speed.

**Comment by ChristianMcFarlaneWork** [comment]
After trying to figure out why the prebuilt binaries were doing a bunch of extra weird stuff to the audio export, I've come to the conclusion that the prebuilt libraries provided above are not working as expected at all.
Basically they just don't seem to work in the current form, manually creating the libraries seems to be the only way to have expected behaviour on Android.

**Comment by gindemit** [comment]
Hi all, sorry for the inconvenience and late reply. I'll have a close look into the problems you mention in the next days. Thanks for trying out the plugin.

**Comment by gindemit** [comment]
Hey guys, I was able to reproduce the issues you mentioned above with System.EntryPointNotFoundException on Android device.
I rebuild the libraries in Android Studio and now the plugin can Save and Load the ogg files on Android.
Please pull the test project before you try to build for Android, I updated the UI on the Main scene.
I have only tested the loading and saving of ogg files. I have not tested the streaming of ogg from file.

Hey @ChristianMcFarlaneWork can you please try out the new release and confirm that there is no "a bunch of extra weird stuff to the audio export". In case you encounter some other issues, please create a new issue here.

Here is the link to the new release which contains only built libraries for the Android OS:
https://github.com/khindemit/unity-vorbis/releases/tag/0.1.1

**Comment by ChristianMcFarlaneWork** [comment]
Hey @khindemit,
Thanks so much for this plugin by the way, It's been a massive help with what we've been working on. The newest libraries you uploaded are working perfectly, no issues with naming or audio speedup on export!
However we've just made a build for iOS and it has the same problem where inspecting the plugin symbols for iOS shows that they have the old name, and the speed issues that the old android build was having.
libVorbisPlugin.dylib symbols:
```
...
000000000000170c T _DecodePcmDataFromFile
0000000000001a5a T _EncodePcmDataToFile
0000000000001a1e T _FreeSamplesArrayNativeMemory
0000000000014846 T __book_maptype1_quantvals
...
```
We're making a new build for iOS now, but it might be worth uploading a new release with the newest version for other developers who might not have the necessary software. 
Cheers, and thankyou again for the plugin!

**Comment by ChristianMcFarlaneWork** [comment]
Cheers,
We've tested out the new release when exporting and it's working exactly as expected, the audio quality is great and we are no longer having issues with the audio speeding up 2x on export, thankyou so much for the help and for the plugin!

After building for iOS we noticed that the library suffers from the same issues that we were encountering with the android libraries. So I've made a separate issue here #2 for that
Cheers, Chris.
